openapi-request-validator: allow nullable on enums

javierblancosp opened the pull request:
The request validation do not allow to have nullable enums. Allow them by transforming openapi specification to AJV
Closes #397

jsdevel commented on packages/openapi-request-validator/test/data-driven.ts:
```diff
@@ -20,7 +20,6 @@ describe(require('../package.json').name, () => {
 
       const validator = new Sut(fixture.validateArgs);
       const error = validator.validate(fixture.request);
-
```
revert this

javierblancosp replied:
Sure sorry :P

jsdevel replied:
still waiting on this to be removed

jsdevel commented on packages/openapi-request-validator/index.ts:
```diff
@@ -20,6 +20,7 @@ export interface OpenAPIRequestValidatorArgs {
   requestBody?: OpenAPIV3.RequestBodyObject;
   schemas?: IJsonSchema[];
   componentSchemas?: IJsonSchema[];
+  ajvOptions: Ajv.Options;
```
not sure how crazy i am about exposing the options like this.  if we decide to go with a different jsonschema validator in the future, we'd have a breaking change to rename this.  is there another way to pass in the options?  and if all this is just to support nullable, shouldn't we just allow it based on what version of OpenAPI is in use?

javierblancosp replied:
mm i see, ok it's a good point. Maybe only with the patch did on https://github.com/kogosoftwarellc/open-api/pull/398/files#diff-32b5cc863d07f4a03bd815a69e54f78cR501 are enough to allow the nullable without exposing the options. What do you think?

jsdevel replied:
Not sure what you mean, but if it's passing in nullable to ajv when v3 schema is detected,then I'd say yes that would be better

TuckerWhitehouse commented:
I have a use case where I need to set `useDefaults: true` in the Ajv options. I was able to monkey patch it, but would greatly benefit from `ajvOptions` support.

jsdevel commented:
@TuckerWhitehouse it's my understanding that the spec dictates that defaults should be enabled with the underlying validator no?

TuckerWhitehouse commented:
@jsdevel I'm not sure I follow. My schema has the following snippet, but when I make a request I get an error that `version` is required -- if I pass `useDefaults` to ajv, the default is applied:
```yaml
  properties:
    version:
	  type: string
      enum: [1.0]
      default: 1.0
```

jsdevel commented:
I believe the OpenAPI schema does require version.  Why would you pass it as a default?  Is that for your API doc?

TuckerWhitehouse commented:
Sorry for the confusion, `version` is a custom prop within the schema.

jsdevel commented:
@TuckerWhitehouse i see.  right now the openapi-default-setter only sets defaults for query string values and headers.  perhaps it should be updated to also populate defaults for request bodies?

i'll state it again: i'm opposed to exposing ajvOptions because it will introduce a breaking change should we decide to go with a different jsonschema validator in the future.  as much as possible, we should try to support the schema.

javierblancosp commented:
@TuckerWhitehouse I don't think it will be a breaking change because is adding a new option. Anyway i'm agree on updating the openapi-default-setter to support the request body population.

Well i removed the parameter :) and added the transformOpenAPIV3Definitions wrapper

jsdevel commented on packages/openapi-request-validator/test/data-driven/accept-requestBody-with-nullable-fields.js:
```diff
@@ -0,0 +1,33 @@
+      }
+    },
+    schemas: null,
+    ajvOptions: {
```
can you remove this?

jsdevel commented on packages/openapi-request-validator/test/data-driven/accept-requestBody-with-nullable-fields-and-enum.js:
```diff
@@ -0,0 +1,34 @@
+      }
+    },
+    schemas: null,
+    ajvOptions: {
```
remove this

jsdevel commented:
@javierblancosp can you squash your commits and update the message so it's more reflective of handling nullable: true instead of adding ajvOptions?

javierblancosp commented:
Sure, i will do it today ;) thanks

jsdevel commented:
sorry @javierblancosp , i merged other PRs and believe i've created conflicts.  can you resolve them?

javierblancosp commented:
Okey, don't worry. Sorry i was away this days, tomorrow i will finalise the pr

jsdevel commented:
thanks @javierblancosp !  this is the next one i merge

javierblancosp commented:
@jsdevel after merge i saw that some test are failing, i tried run master test and i'm getting this error

7 failing

  1) openapi-request-validator
       should accept requestBody with components default:
     AssertionError: expected { Object (status, errors) } to deeply equal undefined
      at Context.<anonymous> (test/data-driven.ts:24:24)

  2) openapi-request-validator
       should accept requestBody with missing readonly required property allOf:
     AssertionError: expected { Object (status, errors) } to deeply equal undefined
      at Context.<anonymous> (test/data-driven.ts:24:24)

  3) openapi-request-validator
       should accept requestBody with missing readonly required property anyOf:
     AssertionError: expected { Object (status, errors) } to deeply equal undefined
      at Context.<anonymous> (test/data-driven.ts:24:24)

  4) openapi-request-validator
       should accept requestBody with missing readonly required property array:
     AssertionError: expected { Object (status, errors) } to deeply equal undefined
      at Context.<anonymous> (test/data-driven.ts:24:24)

  5) openapi-request-validator
       should accept requestBody with missing readonly required property nested refs:
     AssertionError: expected { Object (status, errors) } to deeply equal undefined
      at Context.<anonymous> (test/data-driven.ts:24:24)

  6) openapi-request-validator
       should accept requestBody with missing readonly required property oneOf:
     AssertionError: expected { Object (status, errors) } to deeply equal undefined
      at Context.<anonymous> (test/data-driven.ts:24:24)

  7) openapi-request-validator
       should fail request body when oneOf meets multiple schemas:

      AssertionError: expected { Object (status, errors) } to deeply equal { Object (status, errors) }
      + expected - actual

       {
         "errors": [
           {
      -      "errorCode": "required.openapi.validation"
      -      "location": "body"
      -      "message": "should have required property 'baz'"
      -      "path": "baz"
      -    }
      -    {
      -      "errorCode": "required.openapi.validation"
      -      "location": "body"
      -      "message": "should have required property 'foo'"
      -      "path": "foo"
      -    }
      -    {
      -      "errorCode": "required.openapi.validation"
      -      "location": "body"
      -      "message": "should have required property 'qux'"
      -      "path": "qux"
      -    }
      -    {
             "errorCode": "oneOf.openapi.validation"
             "location": "body"
             "message": "should match exactly one schema in oneOf"
           }

jsdevel commented:
Hmm.  A recent commit must've broken it

javierblancosp commented:
Let me check if i can find the solution

javierblancosp commented:
Hi @jsdevel i think now all is working, also i did the squash ;)

jsdevel commented:
@javierblancosp i'm getting this message:

![conflicts](https://user-images.githubusercontent.com/2605980/56169596-80f6cd00-5f93-11e9-84dd-031794d1a499.png)

javierblancosp commented:
Could you try now, sorry for the mess :S

jsdevel commented:
still the same.  can you rebase your changes instead of merging in master?

jsdevel commented:
@javierblancosp ?

javierblancosp commented:
Sorry ;), i will do something today
